plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -0,0 +1,1333 @@
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1); /* detach */
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 99);
+
+	/* require detach -> attach */
+	locker = fu_device_locker_new_full(device,
+					   (FuDeviceLockerFunc)fu_device_detach,
+					   (FuDeviceLockerFunc)fu_device_attach,
+					   error);
+	if (locker == NULL)
+		return NULL;
+	fu_progress_step_done(progress);
+
+	buf = g_malloc0(size);
+	if (!fu_genesys_usbhub_device_read_flash(self, self->fw_bank_addr[0], buf, size, error))
+		return NULL;
+	fu_progress_step_done(progress);
+
+	return g_bytes_new_take(g_steal_pointer(&buf), size);
+}
+
+static gboolean
+fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	gsize bufsz;
+	guint32 block_size;
+	guint32 sector_size;
+	guint8 static_idx = 0;
+	guint8 dynamic_idx = 0;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GBytes) static_buf = NULL;
+	g_autoptr(GBytes) dynamic_buf = NULL;
+	g_autoptr(GBytes) fw_buf = NULL;
+	g_autoptr(GError) local_error = NULL;
+
+	/* FuUsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_genesys_usbhub_device_parent_class)->setup(device, error)) {
+		g_prefix_error(error, "error setuping device: ");
+		return FALSE;
+	}
+
+	/* [DEBUG] - additional info from device:
+	 * release version: g_usb_device_get_release(usb_device)
+	 */
+
+	/* read standard string descriptors */
+	if (g_usb_device_get_spec(usb_device) >= 0x300) {
+		static_idx = GENESYS_USBHUB_STATIC_TOOL_DESC_IDX_USB_3_0;
+		dynamic_idx = GENESYS_USBHUB_DYNAMIC_TOOL_DESC_IDX_USB_3_0;
+	} else {
+		static_idx = GENESYS_USBHUB_STATIC_TOOL_DESC_IDX_USB_2_0;
+		dynamic_idx = GENESYS_USBHUB_DYNAMIC_TOOL_DESC_IDX_USB_2_0;
+	}
+
+#if G_USB_CHECK_VERSION(0, 3, 8)
+	/*
+	 * Read/parse vendor-specific string descriptors and use that
+	 * data to setup device attributes.
+	 */
+	static_buf =
+	    g_usb_device_get_string_descriptor_bytes_full(usb_device,
+							  static_idx,
+							  G_USB_DEVICE_LANGID_ENGLISH_UNITED_STATES,
+							  64,
+							  error);
+	if (static_buf == NULL) {
+		g_prefix_error(error, "failed to get static tool info from device: ");
+		return FALSE;
+	}
+	if (!fu_genesys_usbhub_device_get_descriptor_data(static_buf,
+							  (guint8 *)&self->static_ts,
+							  sizeof(FuGenesysStaticToolString),
+							  error)) {
+		g_prefix_error(error, "failed to get static tool info from device: ");
+		return FALSE;
+	}
+	if (g_getenv("FWUPD_GENESYS_USBHUB_VERBOSE") != NULL) {
+		fu_common_dump_raw(G_LOG_DOMAIN,
+				   "Static info",
+				   (guint8 *)&self->static_ts,
+				   sizeof(FuGenesysStaticToolString));
+	}
+	if (self->static_ts.tool_string_version != 0xff) {
+		gchar rev[3] = {0};
+		guint64 tmp;
+
+		/* [TODO] move this to the quirk file */
```

Comment:
do you still want to do this? Maybe we just remove the TODO for now as we have hardcoded constants